lib/engine/game/g_1871/game.rb

```diff
@@ -0,0 +1,822 @@
+            end
+          end
+        end
+
+        # A helper function to return the mainline corportaion.
+        def mainline
+          @corporations[0]
+        end
+
+        # A helper function to return the shortline corportaion.
+        def shortline
+          @corporations[1]
+        end
+
+        # The base game pulls all active entities and removes bankrupted
+        # players. We take this but also remove the bank as an active player.
+        def active_players
+          super.reject { |player| player == @union_bank }
+        end
+
+        # Adds the union bank to the end of the player list
+        def add_union_bank_to_players
+          @players << @union_bank unless @players.include?(@union_bank)
+        end
+
+        # This copies the standard code from base for the :shares ability on
+        # companies.
+        #
+        # After it checks for some special 1871 logic:
+        #
+        # - Union Bank: Create the bank player and give it the starting shares
+        #
+        # - PEIR Shares: Create the PEIR shares when someone buys a PEIR private
+        def after_buy_company(player, company, _price)
+          abilities(company, :shares) do |ability|
+            ability.shares.each do |share|
+              share_pool.buy_shares(player, share, exchange: :free)
+            end
+          end
+
+          company.value = 0 if company.sym == 'UB'
+
+          return unless company.id.start_with?('P')
+
+          # Convert PEIR companies to shares
+          share_name = @peir_company_shares[company.id]
+          share = share_by_id(share_name)
+          bundle = Engine::ShareBundle.new(share)
+          share_pool.transfer_shares(bundle, player, price: 0, allow_president_change: false)
+          peir.owner = peir_owner
+        end
+
+        # Returns the player's lowest number PEIR share
+        def lowest_peir_share(player)
+          player.shares_of(peir).min_by(&:index)
+        end
+
+        # Let the Union Bank owner act for the bank in an operating round
+        def acting_for_entity(entity)
+          return entity&.owner unless entity == @union_bank
+
+          bank_company = company_by_id('UB')
+          bank_company.owner
+        end
+
+        # Returns the player that should be the current peir owner
+        def peir_owner
+          # Get all of the share holders
+          share_holders = peir.player_share_holders
+
+          # Get max ownership
+          max_ownership = share_holders.map(&:last).max
+
+          # Get the owners that own the max (tied players)
+          max_owners = share_holders.filter { |_p, o| o == max_ownership }.keys
+
+          # Find the one with the lowest pier share by name
+          max_owners.map { |p| lowest_peir_share(p) }.min_by(&:index).owner
+        end
+
+        # After default caches are made, override the share_by_id and
+        # corporation_by_id method to recognize mainline and shortline
+        def cache_objects
```

Comment:
is this the only way?

Reply:
No, tomorrow I'll experiment with editing abilities to handle this. The idea was that abilities let you exchange for a share of a random company that's set at the start. Instead of editing the abilities during setup I did this, but I can just edit the abilities instead I think, I do that for other reasons anyway.

Reply:
I can change this later if you want but it will be a big change, would rather get this into pre-alpha, get a few spec games going and then be able to safely make the change and know that I didn't break anything.

Reply:
don't you already have tests cases to add specs for?

Reply:
Not yet. I have to do auto passing for the auction next, (no code outside of 1871 for that) and it will break any specs I saved for sure, waiting on that.